local_planner/src/nodes/waypoint_generator.cpp

```diff
@@ -339,13 +339,23 @@ void WaypointGenerator::nextSmoothYaw(float dt) {
 }
 
 void WaypointGenerator::adaptSpeed() {
-  speed_ = planner_info_.cruise_velocity;
+
+  //lowpass filter the speed to get smoother accelerations
+  float lp_filter_constant = 0.1;
+  float last_speed = speed_;
+
+  //at startup parameters are NAN (avoid propagating)
+  if (std::isfinite(planner_info_.cruise_velocity)) {
+    speed_ = planner_info_.cruise_velocity;
+  }else{
+    speed_ = 2.f;
+  }
 
   // If the goal is so close, that the speed-adapted way point would overreach
   float goal_dist = (goal_ - position_).norm();
   if (goal_dist < 1.f) {
-    output_.adapted_goto_position = goal_;
-
+    output_.goto_position = goal_;
+    speed_ = goal_dist;
```

Comment:
Could you explain the rational behind defining the goal distance to speed directly?

Reply:
The speed_  variable defines the distance of the setpoint from the current position of the drone. It is not speed in the general sense though, not in m/s at least. As we use the setpoint as a carrot-stick thing in the firmware, father means faster, closer means slower. So setting the setpoint in goal direction at the distance goal_dist basically puts the setpoint onto the goal.